src/main/java/org/springframework/data/r2dbc/repository/query/ConditionFactory.java

```diff
@@ -0,0 +1,307 @@
+        return SQL.column(dataAccessStrategy.toSql(property.getColumnName()), table);
+    }
+
+    @NotNull
+    @SuppressWarnings("unchecked")
+    private RelationalPersistentEntity<?> getPersistentEntity(PropertyPath propertyPath) {
+        R2dbcConverter converter = dataAccessStrategy.getConverter();
+        return converter.getMappingContext().getRequiredPersistentEntity(propertyPath.getOwningType());
+    }
+
+    @NotNull
+    private BindMarker createBindMarker(ParameterMetadata parameterMetadata) {
+        if (parameterMetadata.getName() != null) {
+            return SQL.bindMarker(":" + parameterMetadata.getName());
+        }
+        return SQL.bindMarker();
+    }
+
+    /**
+     * Applies an {@code UPPERCASE} conversion to the given {@link Expression} in case the underlying {@link Part}
+     * requires ignoring case.
+     *
+     * @param part       method name part (must not be {@literal null})
+     * @param expression expression to be uppercased (must not be {@literal null})
+     * @return uppercased expression or original expression if ignoring case is not strictly required
+     */
+    private Expression upperIfIgnoreCase(Part part, Expression expression) {
+        return upperIfIgnoreCase(part, expression, part.getProperty().getType());
+    }
+
+    /**
+     * Applies an {@code UPPERCASE} conversion to the given {@link Expression} in case the underlying {@link Part}
+     * requires ignoring case.
+     *
+     * @param part           method name part (must not be {@literal null})
+     * @param expression     expression to be uppercased (must not be {@literal null})
+     * @param expressionType type of the given expression (must not be {@literal null})
+     * @return uppercased expression or original expression if ignoring case is not strictly required
+     */
+    private Expression upperIfIgnoreCase(Part part, Expression expression, Class<?> expressionType) {
+        switch (part.shouldIgnoreCase()) {
+            case ALWAYS:
+                Assert.state(canUpperCase(expressionType), "Unable to ignore case of " + expressionType.getName()
+                        + " type, the property '" + part.getProperty().getSegment() + "' must reference a string");
+                return new Upper(expression);
+            case WHEN_POSSIBLE:
+                if (canUpperCase(expressionType)) {
+                    return new Upper(expression);
+                }
+            case NEVER:
+            default:
+                return expression;
+        }
+    }
+
+    private boolean canUpperCase(Class<?> expressionType) {
+        return expressionType == String.class;
+    }
+
+    // TODO: include support of functions in WHERE conditions into spring-data-relational
+    /**
+     * Models the ANSI SQL {@code UPPER} function.
+     * <p>
+     * Results in a rendered function: {@code UPPER(<expression>)}.
+     */
+    private class Upper implements Expression {
```

Comment:
Can we make use of `SimpleFunction` so that we are not required to introduce new types for each function we want to support? It makes sense to expose that functionality as `Functions.upper/lower` in Spring Data Relational.